src/DssSpell.sol

```diff
@@ -110,13 +110,33 @@ contract DssSpell {
     }
 
     modifier limited {
+        require(nextCastTime() == now, "Outside office hours");
+        _;
+    }
+
+    function nextCastTime() public view returns (uint256) {
+        require(eta != 0, "Spell not scheduled");
+        uint256 castTime = now > eta ? now : eta;
+
         if (officeHours) {
-            uint day = (now / 1 days + 3) % 7;
-            require(day < 5, "Can only be cast on a weekday");
-            uint hour = now / 1 hours % 24;
-            require(hour >= 14 && hour < 21, "Outside office hours");
+            uint256 day    = (castTime / 1 days + 3) % 7;
+            uint256 hour   = castTime / 1 hours % 24;
+            uint256 minute = castTime / 1 minutes % 60;
+            uint256 second = castTime % 60;
+
+            if (day >= 5) castTime += 7 days - day * 86400;
+
+            if (hour >= 21) {
+                if (day == 4) castTime += 2 days;
+                castTime += 24 hours - hour * 3600 + 14 hours; // Go to 9am next day
+                castTime -= minute * 60 + second;              // 9am on the hour
+            }
+            else if (hour < 14) {
+                castTime += 14 hours - hour * 3600; // Go to 9am same day
+                castTime -= minute * 60 + second;   // 9am on the hour
+            }
         }
-        _;
+        return castTime;
```

Comment:
Maybe add a `require(castTime <= expiration);` before the `return` since you technically can't `cast` a spell after its expiration.

Reply:
This is an interesting case. The frontend team will be using this function as well, so they'd need to know why it was reverting, my immediate thought was that if it's past the expiration time you could bump `castTime` to `uint(-1)` and it would always fail in the modifier, but maybe a revert is fine here if we give a message that it was expired.

Reply:
We already have the a require for `eta != 0` above--I think it's a reasonable expectation that this function should only return a value if the spell actually _can_ be `cast`. And as you point out, using a `require` allows us to surface an error message; returning `uint(-1)` doesn't.

Reply:
Maybe I'm missing something, but isn't expiration only relevant for `schedule()`? Because this require would fail: `require(now <= expiration, "This contract has expired");`. 

The `eta != 0` check was to ensure that the spell has been scheduled. After that all the pause cares about is that `exec` is done after the `eta`.

So if the spell was scheduled after expiration, then the JS team wouldn't be able to call this since eta would be zero.

Reply:
Yeah, I'm thinking that this function probably does not need to have any knowledge of the expiration, as that's handled in a separate check, this is only for calculating the office hours timestep. Frontend has access to the expiration variable, also, so they can make the check there as well.

Reply:
Yeah those are my thoughts as well.

Reply:
Ahhhh that's true, I thought `expiration` would affect `cast` too, but I was mistaken. Never mind!